### PR TITLE
Fix error message when accessing the component of a frame

### PR DESCRIPTION
**Pull request by eteq**
I recently noticed the following behavior that I found surprising:
```
>>> c_wdata = ICRS(1*u.deg, 2*u.deg)
>>> c_wdata.ra
<Longitude 1.0 deg>
>>> c_nodata = ICRS()
>>> c_nodata.data
ValueError: The frame object "<ICRS Frame>" does not have associated data
>>> c_nodata.ra
AttributeError: 'ICRS' object has no attribute '_rep_cache'
```

I think it's pretty clear that in this context asking ``c.ra`` is a request for data, and the more useful error message is "this frame has no data" rather than an opaque error about not finding the representation cache.

So this PR fixes that by being a bit more careful in the base frame's `__getattr__` about asking for a representation.  So after this PR, the above changes to:
```
>>> c_nodata.ra
ValueError: The frame object "<ICRS Frame>" does not have associated data
```

I think this is more "correcting some possible confusion" than a bug, so I haven't labeled it as a bug, nor added a changelog entry (as it strikes me as more like "documentation").  But @adrn, @astrofrog, or @taldcroft, feel free to say otherwise if you think it should get a changelog entry.

### CODE REVIEW

File: astropy/coordinates/baseframe.py [hunk 1]
```diff
@@ -1037,9 +1037,18 @@ def __getattr__(self, attr):
...
-        rep = self.represent_as(self.representation, in_frame_units=True)
-        val = getattr(rep, self.representation_component_names[attr])
-        return val
+        if self._data is None:
+            if attr in self.representation_component_names.keys():
+                # this raises the "no data" error by design - doing it this way
+                # means we don't have to replicate the error message here
+                self.data
+            else:
+                raise AttributeError("'{0}' object has no attribute '{1}'"
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: Just quickly looking at this before running, but I don't get why you are re-doing the test and exception from lines 1035-1039 here.

Comment (eteq) [inline astropy/coordinates/baseframe.py]: Oh good catch: because I didn't look closely enough at the second ``or`` clause in 1036... will fix.

**Comment by eteq** [comment]
The commit I just pushed up fixed a silly error @taldcroft noticed

**Comment by mhvk** [comment]
Sorry but cannot resist... What if `c.ra` just linked to `c.data.lon`? Wouldn't that be much less confusing still?  See #4903...

**Comment by eteq** [comment]
@mhvk - that would indeed address the situation, but I don't think such a change would be suitable for 1.3, as it would require substantial work (and probably subtle API changes), whereas this is something simple enough to include in 1.3 (and as I said above might even be considered a "doc change" and hence be automatic).

And now that I think about it, this would be a stepa towards that anyway, as ``c.data.lon`` would yield a ``UnitsError`` *not* an ``AttributeError``, so this is in some sense prepping for that change anyway.

**Comment by mhvk** [comment]
I didn't mean to imply that this depended on resolving #4903... 

I just restarted a few travis builds that had hung (there do seem to be quite a few network problems lately, including conda complaining that its server is at capacity).

**Comment by eteq** [comment]
OK, so do you think this looks good for now, @mhvk? (particularly for 1.2, as it's mostly doc)

**Comment by astrofrog** [comment]
👍  to backporting to 1.2

**Comment by astrofrog** [comment]
Having said that, it either needs a changelog entry, or the ``no-changelog-entry-needed`` label 😁

**Comment by mhvk** [comment]
Yes, this looks good. So, adding the relevant label and merging!

**Pull request merged**